### PR TITLE
Adds efficient view/layer push/pull configurations.

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

**Comment by invisig0th** [comment]
Still need to get perms figured out.

File: synapse/cortex.py [hunk 1]
```diff
@@ -2771,6 +2812,146 @@ async def _initCoreLayers(self):
...
+    async def _pushBulkEdits(self, layr0, layr1, pdef):
+
+        iden = pdef.get('iden')
+        user = pdef.get('user')
+
+        gvar = f'push:{iden}'
+
+        async with await s_base.Base.anit() as base:
+
+            queue = s_queue.Queue(maxsize=10000)
```
Comment (jnwatson) [inline synapse/cortex.py:2921]: there's no need to make the queue this big.

File: synapse/lib/cell.py [hunk 2]
```diff
@@ -904,13 +905,74 @@ async def _setAhaActive(self):
         except Exception as e: # pragma: no cover
             logger.warning(f'_setAhaActive failed: {e}')
 
+    async def addActiveCoro(self, func, iden=None):
```
Comment (jnwatson) [inline synapse/lib/cell.py:908]: Pretty please add a one line description.

File: synapse/cortex.py [hunk 3]
```diff
@@ -2771,6 +2812,146 @@ async def _initCoreLayers(self):
...
+        await view.info.set('pulls', pulls)
+        await self.delActiveCoro(pulliden)
+
+    async def runLayrPush(self, layr, pdef):
+        url = pdef.get('url')
+        iden = pdef.get('iden')
+        # push() will refire as needed
+        async def push():
+            async with await self.boss.promote(f'layer push: {layr.iden} {iden}', self.auth.rootuser) as task:
+                async with await s_telepath.openurl(url) as proxy:
```
Comment (jnwatson) [inline synapse/cortex.py:2898]: What if we get a disconnect?

File: synapse/cortex.py [hunk 4]
```diff
@@ -2771,6 +2812,146 @@ async def _initCoreLayers(self):
...
+            return
+
+        await view.info.set('pulls', pulls)
+        await self.delActiveCoro(pulliden)
+
+    async def runLayrPush(self, layr, pdef):
+        url = pdef.get('url')
+        iden = pdef.get('iden')
+        # push() will refire as needed
+        async def push():
```
Comment (jnwatson) [inline synapse/cortex.py:2896]: pep8: newline before function

File: synapse/cortex.py [hunk 5]
```diff
@@ -2771,6 +2812,146 @@ async def _initCoreLayers(self):
...
+            async with await self.boss.promote(f'layer push: {layr.iden} {iden}', self.auth.rootuser) as task:
+                async with await s_telepath.openurl(url) as proxy:
+                    await self._pushBulkEdits(layr, proxy, pdef)
+        await self.addActiveCoro(push, iden=iden)
+
+    async def runViewPull(self, view, pdef):
+        url = pdef.get('url')
+        iden = pdef.get('iden')
+        # pull() will refire as needed
+        async def pull():
```
Comment (jnwatson) [inline synapse/cortex.py:2906]: pep8: newline before function

File: synapse/cortex.py [hunk 6]
```diff
@@ -2771,6 +2812,146 @@ async def _initCoreLayers(self):
...
+            async with await self.boss.promote(f'view pull: {view.iden} {iden}', self.auth.rootuser) as task:
```
Comment (jnwatson) [inline synapse/cortex.py:2907]: unused variable: task

File: synapse/cortex.py [hunk 7]
```diff
@@ -2771,6 +2812,146 @@ async def _initCoreLayers(self):
...
+
+        iden = pdef.get('iden')
+        user = pdef.get('user')
+
+        gvar = f'push:{iden}'
+
+        async with await s_base.Base.anit() as base:
+
+            queue = s_queue.Queue(maxsize=10000)
+            async def fill():
```
Comment (jnwatson) [inline synapse/cortex.py:2922]: newline before function

File: synapse/lib/cell.py [hunk 8]
```diff
@@ -904,13 +905,74 @@ async def _setAhaActive(self):
...
+            except Exception as e:
+                logger.warning(f'delActiveCoro Task: {e}')
+
+    async def _fireActiveCoros(self):
+        for iden, cdef in self.activecoros.items():
+            await self._fireActiveCoro(iden, cdef)
+
+    async def _fireActiveCoro(self, iden, cdef):
+        func = cdef.get('func')
+        async def wrap():
```
Comment (jnwatson) [inline synapse/lib/cell.py:943]: empty line before function

File: synapse/lib/cell.py [hunk 9]
```diff
@@ -904,13 +905,74 @@ async def _setAhaActive(self):
...
+                except asyncio.CancelledError:
+                    raise
+                except Exception as e:
+                    logger.warning(f'activeCoro Error: {func} {e}')
+                    await asyncio.sleep(1)
+
+        cdef['task'] = self.schedCoro(wrap())
+
+    async def _killActiveCoros(self):
+        for iden, cdef in self.activecoros.items():
```
Comment (jnwatson) [inline synapse/lib/cell.py:956]: unused variable iden (_iden works)

File: synapse/lib/stormtypes.py [hunk 10]
```diff
@@ -3213,11 +3223,70 @@ def getObjLocals(self):
...
+        layriden = self.valu.get('iden')
+
+        if not self.runt.isAdmin(gateiden=layriden):
+            mesg = '$layer.addPush() requires admin privs on the layer.'
+            raise s_exc.AuthDeny(mesg=mesg)
+
+        scheme = url.split('://')[0]
+        self.runt.confirm(('lib', 'telepath', 'open', scheme))
+
+        async with await s_telepath.openurl(url) as proxy:
```
Comment (jnwatson) [inline synapse/lib/stormtypes.py:3258]: unused variable proxy

File: synapse/lib/stormtypes.py [hunk 11]
```diff
@@ -3436,11 +3513,81 @@ def getObjLocals(self):
...
+            raise s_exc.AuthDeny(mesg=mesg)
+
+        if not self.runt.isAdmin(gateiden=layriden):
+            mesg = '$view.addPull() requires admin privs on the top layer.'
+            raise s_exc.AuthDeny(mesg=mesg)
+
+        scheme = url.split('://')[0]
+        self.runt.confirm(('lib', 'telepath', 'open', scheme))
+
+        async with await s_telepath.openurl(url) as proxy:
```
Comment (jnwatson) [inline synapse/lib/stormtypes.py:3553]: unused variable proxy

**Review by jnwatson — APPROVED** [review]

File: synapse/cortex.py (hunk 3 above)
Comment (jnwatson) [reply synapse/cortex.py:2898]: nm I see it.

File: synapse/cortex.py (hunk 1 above)
Comment (invisig0th) [reply synapse/cortex.py:2921]: We'll be consuming in chunks of up to 1000 queue entries at a time and i want to ensure constant/smooth consumption.  These will be nodeedits() tuples and have a very small footprint, so i see no harm in allowing up to 10k.  This approach allows large perf/efficiency gains when back-filling data ( especially into a cluster by keeping transaction counts lower ) and costs very little.

File: synapse/lib/cell.py (hunk 9 above)
Comment (invisig0th) [reply synapse/lib/cell.py:956]: Actually, this is a bug!  Looks like killActiveCoros() isn't actually removing them.  Will fix.

**Review by jnwatson — APPROVED** [review]

**Pull request merged**